src/platform_linux/platform_inline.h

```diff
@@ -227,64 +227,71 @@ platform_status_to_string(const platform_status status)
 }
 
 /* Default output file handles for different logging interfaces */
-#define PLATFORM_DEFAULT_LOG_HANDLE stdout
-#define PLATFORM_ERR_LOG_HANDLE     stderr
-#define PLATFORM_CR                 "\r"
-
-#define platform_open_log_stream()                                             \
-   char                  *bp;                                                  \
-   size_t                 size;                                                \
-   platform_stream_handle stream;                                              \
-   stream = open_memstream(&bp, &size);
-
-#define platform_close_log_stream(h)                                           \
-   fclose(stream);                                                             \
-   fputs(bp, h);                                                               \
-   fflush(h);                                                                  \
-   platform_free(NULL, bp);
+#define PLATFORM_CR "\r"
+
+static inline platform_status
+platform_open_log_stream(platform_stream_handle *stream)
+{
+   ZERO_CONTENTS(stream);
+   stream->stream = open_memstream(&stream->str, &stream->size);
+   if (stream->stream == NULL) {
+      return STATUS_NO_MEMORY;
+   }
+   return STATUS_OK;
+}
+
+static inline void
+platform_flush_log_stream(platform_stream_handle *stream)
+{
+   fflush(stream->stream);
+}
+
+static inline void
+platform_close_log_stream(platform_stream_handle *stream,
+                          platform_log_handle    *log_handle)
+{
+   fclose(stream->stream);
+   fputs(stream->str, log_handle);
+   fflush(log_handle);
+   platform_free_from_heap(NULL, stream->str);
+}
+
+static inline platform_log_handle *
+platform_log_stream_to_log_handle(platform_stream_handle *stream)
+{
+   return stream->stream;
+}
+
+static inline char *
+platform_log_stream_to_string(platform_stream_handle *stream)
+{
+   platform_flush_log_stream(stream);
+   return stream->str;
+}
 
 /*
  * platform_log_stream should be called between platform_open_log_stream and
  * platform_close_log_stream.
  */
-#define platform_log_stream(...)                                               \
-   do {                                                                        \
-      fprintf(stream, __VA_ARGS__);                                            \
-   } while (0)
-
-#define platform_log(...)                                                      \
+#define platform_log_stream(_stream, ...)                                      \
    do {                                                                        \
-      fprintf(Platform_stdout_fh, __VA_ARGS__);                                \
-      fflush(Platform_stdout_fh);                                              \
+      fprintf(_stream->stream, __VA_ARGS__);                                   \
```

Comment:
Suggestion: What you have also works, but ... now that you have tried to make all outputs go to log-handles, and given that you have also provided this `platform_log_stream_to_log_handle()` interfaces, seems like we can rework this line to do:

```
   platform_log(platform_log_stream_to_log_handle(_stream), __VA_ARGS__);
```

Basically, I am wondering why not have the `log_stream()` also invoke `platform_log_handle()`, so everything is funneled thru the latter?

I think from layering angle that seems right, but wonder if it will make code readability too too hard?

I didn't think it would make the calling sequence too hard to de-mangle, so am suggesting you consider bringing this in.

Reply:
This is fine with me.

Really it may make sense to get rid of this function altogether now and just pass the `log_handle` obtained from the stream to `platform_log`. I'll create an issue for that I think.